File: Game_Of_Life_Python/Organisms/animal.py
from abc import ABC, abstractmethod
from Organisms.organism import Organism
import random
from enum import Enum

class MoveDirection(Enum):
    LEFT = 1
    RIGHT = 2
    TOP = 3
    BOTTOM = 4

class CollisionTypes(Enum):
    NONE = "None"
    MULTIPLICATION = "Multiplication"
    FIGHT = "Fight"

class Animal(Organism, ABC):

    MAX_ANIMAL_AMOUNT = 5
    MIN_STRENGTH_TURTLE_REFLECT = 4

    def __init__(self, strength, initiative, name, character, row, column, image_path):
        super().__init__(strength, initiative, name, character, row, column, image_path)

    @abstractmethod
    def action(self, grid_board):
        pass

    @abstractmethod
    def collision(self, grid_board, organisms, current_index):
        pass

    @abstractmethod
    def get_static_counter(self):
        pass

    @abstractmethod
    def increase_static_counter(self):
        pass

    @abstractmethod
    def decrease_static_counter(self):
        pass

    def default_action_animal(self, grid_board):

        self.age += 1

        self.previous_row = self.row
        self.previous_column = self.column

        direction = MoveDirection(random.randint(1, 4))

        # Check if the movement in the selected direction will keep the organism within the board bounds
        while True:
            if self.row == 0 and direction == MoveDirection.TOP:
                direction = MoveDirection(random.choice(
                    [MoveDirection.LEFT, MoveDirection.RIGHT, MoveDirection.BOTTOM]))
            elif self.row == 9 and direction == MoveDirection.BOTTOM:
                direction = MoveDirection(random.choice(
                    [MoveDirection.LEFT, MoveDirection.RIGHT, MoveDirection.TOP]))
            elif self.column == 0 and direction == MoveDirection.LEFT:
                direction = MoveDirection(random.choice(
                    [MoveDirection.TOP, MoveDirection.RIGHT, MoveDirection.BOTTOM]))
            elif self.column == 9 and direction == MoveDirection.RIGHT:
                direction = MoveDirection(random.choice(
                    [MoveDirection.TOP, MoveDirection.LEFT, MoveDirection.BOTTOM]))
            else:
                break

        grid_board[self.row][self.column] = 'e'
        name = self.name
        if name == "cyber_sheep":
            name = "cyber"
        self.print_to_journal(
            f"{name} from ({self.row}, {self.column}) to ")

        # Perform movement based on the selected direction
        if direction == MoveDirection.LEFT:
            self.organism_go_left()
        elif direction == MoveDirection.TOP:
            self.organism_go_top()
        elif direction == MoveDirection.RIGHT:
            self.organism_go_right()
        elif direction == MoveDirection.BOTTOM:
            self.organism_go_bottom()

        grid_board[self.row][self.column] = self.character
        self.print_to_journal(f"to ({self.row}, {self.column})\n")

    def default_collision_animal(self, grid_board, organisms, current_index):
        
        for organism_index, organism in enumerate(organisms):

            if organism.get_position_row() == self.row and organism.get_position_column() == self.column and organism_index != current_index:

                # Some special plant cases
                if organism.get_name() == "Guarana" or organism.get_name() == "Belladonna" or organism.get_name() == "Sosnowsky_hogweed":
                    # self.print_to_journal(f"{self.character} vs {organism.get_character()} at ({self.row}, {self.column})\n")
                    return CollisionTypes("None"), None

                # Multiplication case
                elif self.character == organism.get_character():
                    
                    self.row = self.previous_row
                    self.column = self.previous_column
                    grid_board[self.row][self.column] = self.character
                    if self.get_static_counter() > Animal.MAX_ANIMAL_AMOUNT:
                        self.print_to_journal(f"""{self.character} vs {organism.get_character()} -> cant be more than {Animal.MAX_ANIMAL_AMOUNT}\n""")
                        return CollisionTypes("None"), None

                    else:
                        return CollisionTypes("Multiplication"), self.type(self.name, self.row, self.column)

                # Turtle collision
                elif organism.get_name() == 'turtle' and self.strength <= Animal.MIN_STRENGTH_TURTLE_REFLECT:
                    
                    self.row = self.previous_row
                    self.column = self.previous_column
                    grid_board[self.row][self.column] = self.character
                    self.print_to_journal(f"""{self.character} vs {organism.get_character()} -> {self.character} back to ({self.row}, {self.column})\n""")
                    return CollisionTypes("None"), None

                # Fight case
                else:

                    if self.strength >= organism.get_strength():
                        
                        self.print_to_journal(f"""{self.character} vs {organism.get_character()} -> {self.character} wins at ({self.row}, {self.column})\n""")
                        grid_board[self.row][self.column] = self.character
                        return CollisionTypes("Fight"), organism_index

                    else:

                        self.print_to_journal(f"""{self.character} vs {organism.get_character()} -> {organism.get_character()} wins at ({self.row}, {self.column})\n""")
                        grid_board[self.row][self.column] = organism.get_character()
                        return CollisionTypes("Fight"), current_index


        grid_board[self.row][self.column] = self.character
        return CollisionTypes("None"), None

File: GameOfLife_Cpp/source/src/animal.cpp
#include "animal.hpp"
#include <cassert>
#include <iostream>

Animal::Animal(int strength, int initiative, int age, std::string name, char ch, int r, int c, std::string image_name, OrganismType t)
    : Organism(strength, initiative, age, name, ch, r, c, image_name, t)
{
    
}

ActionResult Animal::default_action_animal(std::vector<std::vector<char>>& grid_board)
{
    this->previous_row = this->row;
    this->previous_column = this->column;

    Direction direction = static_cast<Direction>((rand()%4)+1);
    while (true)
    {
        if ((this->row == 0 && direction == Direction::TOP)
            || (this->row == BOARD_HEIGHT - 1 && direction == Direction::BOTTOM)
            || (this->column == 0 && direction == Direction::LEFT)
            || (this->column == BOARD_WIDTH - 1 && direction == Direction::RIGHT))
            direction = static_cast<Direction>((rand()%4)+1);
        else
            break;
    }

    switch (direction)
    {
        case Direction::LEFT:
            this->move_left();
            break;
        case Direction::TOP:
            this->move_top();
            break;
        case Direction::RIGHT:
            this->move_right();
            break;
        case Direction::BOTTOM:
            this->move_bottom();
            break;
        default:
            std::cout << "Invalid direction at default_animal_action(...)\n";
            break;
    }

    this->move_message();
    this->default_grid_update(grid_board);
    return ActionResult(ActionType::MOVE);
}

CollisionResult Animal::default_collision_animal(std::vector<std::vector<char>>& grid_board, std::vector<Organism*>& organisms, int current_index)
{
    int index = 0;
    std::vector<int> indexes;
    for (auto& organism : organisms)
    {
        // We found collision with another organism
        if (organism->get_row() == this->row && organism->get_column() == this->column && index != current_index)
        {
            // Multiplication case
            if (organism->get_type() == this->get_type())
            {
                assert(previous_row >= 0 && previous_row < BOARD_HEIGHT && previous_column >= 0 && previous_column < BOARD_WIDTH);
                std::cout << this->get_name() << " multiplication at (" << this->row << ", " << column << ") " 
                << this->get_name() << " goes back to (" << this->previous_row << ", " << this->previous_column << ")\n";
                int original_row = row;
                int original_col = column;
                row = previous_row;
                column = previous_column;
                assert(row >= 0 && row < BOARD_HEIGHT && column >= 0 && column < BOARD_WIDTH);
                grid_board[row][column] = this->get_character();

                // Check for max amount
                if (this->get_static_counter() >= MAX_ORGANISM_AMOUNT)
                {
                    std::cout << "Cannot add more " << this->get_name() << ", we arleady have " << this->get_static_counter() << " of them on the board...\n";
                    return CollisionResult(CollisionType::NONE);
                }

                return CollisionResult(CollisionType::MULTIPLICATION, Point(original_row, original_col), this->get_type());
            }
            // Turtle collision case
            else if (organism->get_type() == OrganismType::TURTLE && this->get_strength() < 5)
            {
                std::cout << organism->get_name() << " reflects the attack at (" << this->row << ", " << column << ") " 
                << this->get_name() << " goes back to (" << previous_row << ", " << previous_column << ")\n";
                grid_board[row][column] = 't';
                row = previous_row;
                column = previous_column;
                grid_board[row][column] = this->get_character();
                return CollisionResult(CollisionType::NONE);
            }
            // Fight case
            else if (organism->get_type() != this->get_type())
            {
                // Poison plants case
                if ((organism->get_type() == OrganismType::SOSNOWSKY_HOGWEED && this->get_type() != OrganismType::CYBER_SHEEP) 
                || organism->get_type() == OrganismType::BELLADONNA)
                {
                    std::cout << get_name() << " eats " << organism->get_name() << " and " 
                    << organism->get_name() << " eats " << get_name() << " at (" << row << ", " << column << ")\n";
                    indexes.push_back(current_index);
                    indexes.push_back(index);
                    return CollisionResult(CollisionType::FIGHT, indexes);
                }
                
                if (this->get_strength() >= organism->get_strength())
                {
                    std::cout << this->get_name() << " wins with " << organism->get_name() << " at (" << this->row << ", " << column << ")\n";
                    indexes.push_back(index);
                    grid_board[row][column] = this->get_character();
                    
                    // If we eat guarana we get +3 to strength permanently
                    if (organism->get_type() == OrganismType::GUARANA)
                    {
                        this->get_default_strength() += 3;
                        this->get_strength() += 3;
                        std::cout << this->get_name() << " +3 to strength, current strength = " << this->get_strength() << "...\n";
                    }
                }
                else 
                {
                    std::cout << organism->get_name() << " wins with " << this->get_name() << " at (" << this->row << ", " << column << ")\n";
                    indexes.push_back(current_index);
                    grid_board[row][column] = organism->get_character();
                }
                return CollisionResult(CollisionType::FIGHT, indexes);
            }
        }
        index++;
    }
    return CollisionResult(CollisionType::NONE);
}